tls/s2n_handshake_io.c

```diff
@@ -694,7 +694,16 @@ static int s2n_advance_message(struct s2n_connection *conn)
 }
 
 int s2n_generate_new_client_session_id(struct s2n_connection *conn)
+    CONTRACT_REQUIRES(__CPROVER_is_fresh(conn, sizeof(*conn)) || (conn == NULL))
+    CONTRACT_ASSIGNS_ERR(conn->session_id_len, conn->session_id)
+    CONTRACT_ENSURES(S2N_IMPLIES(conn == NULL, __CPROVER_return_value == S2N_FAILURE))
+    CONTRACT_ENSURES(S2N_IMPLIES(__CPROVER_return_value == S2N_SUCCESS && conn->mode == S2N_SERVER, conn->session_id_len == S2N_TLS_SESSION_ID_MAX_LEN))
+    CONTRACT_ENSURES(S2N_IMPLIES(conn != NULL && conn->mode != S2N_SERVER, __CPROVER_return_value == S2N_SUCCESS))
+    /* CONTRACT_ENSURES(S2N_IMPLIES(conn->mode != S2N_SERVER, *conn == __CPROVER_old(*conn)))              */ /* Check does not terminate */
+    /* CONTRACT_ENSURES(S2N_IMPLIES(__CPROVER_return_value != S2N_SUCCESS, *conn == __CPROVER_old(*conn))) */ /* Check does not terminate */
```

Comment:
The assigns clause already proves that the only memory location that may change is `conn->session_id_len`, so why do we need to re-state again that `*conn == __CPROVER_old(*conn)`? Perhaps, we could add the following contracts
```c
CONTRACT_ENSURES(S2N_IMPLIES(conn->mode != S2N_SERVER, conn->session_id_len == __CPROVER_old(conn->session_id_len))) 
CONTRACT_ENSURES(S2N_IMPLIES(__CPROVER_return_value != S2N_SUCCESS, conn->session_id_len == __CPROVER_old(conn->session_id_len)))
```
What do you think? Also, I don't think that we current have support for struct members in `__CPROVER_old`, so in any case, we could create a GitHub issue to add this check later.

Reply:
Agree ... in fact what I tried to do first is what Felipe suggested, but __CPROVER_old on a struct field wasn't supported then (is it now?) So, I tried instead to use the whole struct, which works syntactically, the analysis just doesn't terminate.  

I'm happy to delete these lines provided we log issues on those two problems (these lines kind of existed to help me remember them).